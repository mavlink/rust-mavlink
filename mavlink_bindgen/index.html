<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `mavlink_bindgen` crate."><title>mavlink_bindgen - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-528d32ef.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="mavlink_bindgen" data-themes="" data-resource-suffix="" data-rustdoc-version="1.91.0-nightly (6ba0ce409 2025-08-21)" data-channel="nightly" data-search-js="search-e4956047.js" data-stringdex-js="stringdex-0e748618.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-1b37d467.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-b6a35c53.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Crate mavlink_bindgen</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../mavlink_bindgen/index.html">mavlink_<wbr>bindgen</a><span class="version">0.15.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#reexports">Crate Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>mavlink_<wbr>bindgen</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/mavlink_bindgen/lib.rs.html#1-210">Source</a> </span></div><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><dl class="item-table reexports"><dt id="reexport.BindGenError"><code>pub use crate::error::<a class="enum" href="error/enum.BindGenError.html" title="enum mavlink_bindgen::error::BindGenError">BindGenError</a>;</code></dt></dl><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><dl class="item-table"><dt><a class="mod" href="binder/index.html" title="mod mavlink_bindgen::binder">binder</a></dt><dt><a class="mod" href="error/index.html" title="mod mavlink_bindgen::error">error</a></dt><dt><a class="mod" href="parser/index.html" title="mod mavlink_bindgen::parser">parser</a></dt></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.GeneratedBinding.html" title="struct mavlink_bindgen::GeneratedBinding">Generated<wbr>Binding</a></dt><dt><a class="struct" href="struct.GeneratedBindings.html" title="struct mavlink_bindgen::GeneratedBindings">Generated<wbr>Bindings</a></dt></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.XmlDefinitions.html" title="enum mavlink_bindgen::XmlDefinitions">XmlDefinitions</a></dt><dd>Specifies the source(s) of MAVLink XML definition files used for generating
Rust MAVLink dialect bindings.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.emit_cargo_build_messages.html" title="fn mavlink_bindgen::emit_cargo_build_messages">emit_<wbr>cargo_<wbr>build_<wbr>messages</a></dt><dd>Prints definitions for cargo that describe which files the generated code depends on, indicating when it has to be regenerated.</dd><dt><a class="fn" href="fn.format_generated_code.html" title="fn mavlink_bindgen::format_generated_code">format_<wbr>generated_<wbr>code</a></dt><dd>Formats generated code using <code>rustfmt</code>.</dd><dt><a class="fn" href="fn.generate.html" title="fn mavlink_bindgen::generate">generate</a></dt><dd>Generate Rust MAVLink dialect binding for dialects present in the given <code>xml_definitions</code>
into <code>destination_dir</code>.</dd></dl></section></div></main></body></html>