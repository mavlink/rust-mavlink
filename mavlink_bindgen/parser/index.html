<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `parser` mod in crate `mavlink_bindgen`."><title>mavlink_bindgen::parser - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-069232aa.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="mavlink_bindgen" data-themes="" data-resource-suffix="" data-rustdoc-version="1.91.0-nightly (425a9c0a0 2025-08-17)" data-channel="nightly" data-search-js="search-b069538f.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-68b7e25d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-eebb9057.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../mavlink_bindgen/index.html">mavlink_<wbr>bindgen</a><span class="version">0.15.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module parser</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate mavlink_<wbr>bindgen</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">mavlink_bindgen</a></div><h1>Module <span>parser</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/mavlink_bindgen/parser.rs.html#1-1675">Source</a> </span></div><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.MavEnum.html" title="struct mavlink_bindgen::parser::MavEnum">MavEnum</a></dt><dt><a class="struct" href="struct.MavEnumEntry.html" title="struct mavlink_bindgen::parser::MavEnumEntry">MavEnum<wbr>Entry</a></dt><dt><a class="struct" href="struct.MavField.html" title="struct mavlink_bindgen::parser::MavField">MavField</a></dt><dt><a class="struct" href="struct.MavMessage.html" title="struct mavlink_bindgen::parser::MavMessage">MavMessage</a></dt><dt><a class="struct" href="struct.MavProfile.html" title="struct mavlink_bindgen::parser::MavProfile">MavProfile</a></dt></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.MavType.html" title="enum mavlink_bindgen::parser::MavType">MavType</a></dt><dt><a class="enum" href="enum.MavXmlElement.html" title="enum mavlink_bindgen::parser::MavXmlElement">MavXml<wbr>Element</a></dt></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.extra_crc.html" title="fn mavlink_bindgen::parser::extra_crc">extra_<wbr>crc</a></dt><dd>CRC operates over names of the message and names of its fields
Hence we have to preserve the original uppercase names delimited with an underscore
For field names, we replace “type” with “mavtype” to make it rust compatible (this is
needed for generating sensible rust code), but for calculating crc function we have to
use the original name “type”</dd><dt><a class="fn" href="fn.generate.html" title="fn mavlink_bindgen::parser::generate">generate</a></dt><dd>Generate protobuf represenation of mavlink message set
Generate rust representation of mavlink message set with appropriate conversion methods</dd><dt><a class="fn" href="fn.parse_profile.html" title="fn mavlink_bindgen::parser::parse_profile">parse_<wbr>profile</a></dt></dl></section></div></main></body></html>