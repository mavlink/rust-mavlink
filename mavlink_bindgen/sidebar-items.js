window.SIDEBAR_ITEMS = {"enum":["XmlDefinitions"],"fn":["emit_cargo_build_messages","format_generated_code","generate"],"mod":["binder","error","parser"],"struct":["GeneratedBinding","GeneratedBindings"]};