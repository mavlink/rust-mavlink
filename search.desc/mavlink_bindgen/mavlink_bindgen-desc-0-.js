searchState.loadedDescShard("mavlink_bindgen", 0, "Prints definitions for cargo that describe which files the …\nFormats generated code using <code>rustfmt</code>.\nReturns the argument unchanged.\nReturns the argument unchanged.\nGenerate Rust MAVLink dialect binding for dialects present …\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nRepresents a failure to create a Rust file for the …\nRepresents a failure to read a MAVLink definition file.\nRepresents a failure to read the MAVLink definitions …\nRepresents a failure to read a directory entry in the …\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCompare two MavTypes\nCRC operates over names of the message and names of its …\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nGenerate protobuf represenation of mavlink message set …\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nIf contains Some, the string represents the primitive type …\nUsed for crc calculation\nReturn rust equivalent of the primitive type of a MavType. …\nEmit reader of a given type\nReturn rust equivalent of a given Mavtype Used for …\nEmit writer of a given type")