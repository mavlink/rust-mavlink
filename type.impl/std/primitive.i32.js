(function() {
    var type_impls = Object.fromEntries([["mio_serial",[]],["nix",[["<details class=\"toggle implementors-toggle\" open><summary><section id=\"impl-ErrnoSentinel-for-i32\" class=\"impl\"><a class=\"src rightside\" href=\"src/nix/errno.rs.html#151-155\">Source</a><a href=\"#impl-ErrnoSentinel-for-i32\" class=\"anchor\">ยง</a><h3 class=\"code-header\">impl <a class=\"trait\" href=\"nix/errno/trait.ErrnoSentinel.html\" title=\"trait nix::errno::ErrnoSentinel\">ErrnoSentinel</a> for <a class=\"primitive\" href=\"https://doc.rust-lang.org/nightly/std/primitive.i32.html\">i32</a></h3></section></summary><div class=\"impl-items\"><section id=\"method.sentinel\" class=\"method trait-impl\"><a class=\"src rightside\" href=\"src/nix/errno.rs.html#152-154\">Source</a><a href=\"#method.sentinel\" class=\"anchor\">ยง</a><h4 class=\"code-header\">fn <a href=\"nix/errno/trait.ErrnoSentinel.html#tymethod.sentinel\" class=\"fn\">sentinel</a>() -&gt; Self</h4></section></div></details>","ErrnoSentinel","nix::pty::SessionId"]]],["serde",[]],["serde_core",[]],["tokio",[]]]);
    if (window.register_type_impls) {
        window.register_type_impls(type_impls);
    } else {
        window.pending_type_impls = type_impls;
    }
})()
//{"start":55,"fragment_lengths":[17,966,13,18,13]}