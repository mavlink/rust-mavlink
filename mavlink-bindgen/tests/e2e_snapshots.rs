use std::fs;
use std::path::PathBuf;

use insta::{self, assert_snapshot, glob};
use mavlink_bindgen::{format_generated_code, generate, XmlDefinitions};
use tempfile::TempDir;

fn definitions_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("definitions")
}

fn run_snapshot(def_file: &str) {
    let defs = definitions_dir();
    let tmp = TempDir::new().expect("tmp dir");
    let out_dir = tmp.path();

    let xml = defs.join(def_file);
    let result = generate(XmlDefinitions::Files(vec![xml]), out_dir).expect("generate ok");

    format_generated_code(&result);

    glob!(out_dir, "**/*.rs", |path| {
        let contents = fs::read_to_string(path).expect("read generated file");
        assert_snapshot!(def_file, contents);
    })
}

#[test]
fn snapshot_heartbeat() {
    run_snapshot("heartbeat.xml");
}

#[test]
fn snapshot_parameters() {
    run_snapshot("parameters.xml");
}

#[test]
fn snapshot_deprecated() {
    run_snapshot("deprecated.xml");
}
