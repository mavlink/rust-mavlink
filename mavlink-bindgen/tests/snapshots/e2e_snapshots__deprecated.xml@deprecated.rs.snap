---
source: mavlink-bindgen/tests/e2e_snapshots.rs
assertion_line: 26
expression: contents
---
#![doc = "MAVLink deprecated dialect."]
#![doc = ""]
#![doc = "This file was automatically generated, do not edit."]
#![allow(deprecated)]
#[cfg(feature = "arbitrary")]
use arbitrary::Arbitrary;
#[allow(unused_imports)]
use bitflags::bitflags;
use mavlink_core::{bytes::Bytes, bytes_mut::BytesMut, MavlinkVersion, Message, MessageData};
#[allow(unused_imports)]
use num_derive::FromPrimitive;
#[allow(unused_imports)]
use num_derive::ToPrimitive;
#[allow(unused_imports)]
use num_traits::FromPrimitive;
#[allow(unused_imports)]
use num_traits::ToPrimitive;
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};
#[cfg(feature = "ts")]
use ts_rs::TS;
#[cfg_attr(feature = "ts", derive(TS))]
#[cfg_attr(feature = "ts", ts(export))]
#[derive(Debug, Copy, Clone, PartialEq, FromPrimitive, ToPrimitive)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type"))]
#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
#[repr(u32)]
pub enum MavFrame {
    #[deprecated = "Use MAV_FRAME_GLOBAL in COMMAND_INT (and elsewhere) as a synonymous replacement. See `MAV_FRAME_GLOBAL` (Deprecated since 2024-03)"]
    #[doc = "Global (WGS84) coordinate frame (scaled) + altitude relative to mean sea level (MSL)."]
    MAV_FRAME_GLOBAL_INT = 5,
}
impl MavFrame {
    pub const DEFAULT: Self = Self::MAV_FRAME_GLOBAL_INT;
}
impl Default for MavFrame {
    fn default() -> Self {
        Self::DEFAULT
    }
}
#[cfg_attr(feature = "ts", derive(TS))]
#[cfg_attr(feature = "ts", ts(export))]
#[derive(Debug, Copy, Clone, PartialEq, FromPrimitive, ToPrimitive)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type"))]
#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
#[repr(u32)]
#[deprecated = " See `GIMBAL_MANAGER_FLAGS` (Deprecated since 2020-01)"]
#[doc = "Enumeration of possible mount operation modes. This message is used by obsolete/deprecated gimbal messages."]
pub enum MavMountMode {
    #[doc = "Load and keep safe position (Roll,Pitch,Yaw) from permanent memory and stop stabilization"]
    MAV_MOUNT_MODE_RETRACT = 0,
}
impl MavMountMode {
    pub const DEFAULT: Self = Self::MAV_MOUNT_MODE_RETRACT;
}
impl Default for MavMountMode {
    fn default() -> Self {
        Self::DEFAULT
    }
}
#[deprecated = "To be removed / merged with TIMESYNC. See `TIMESYNC` (Deprecated since 2011-08)"]
#[doc = "A ping message either requesting or responding to a ping. This allows to measure the system latencies, including serial port, radio modem and UDP connections. The ping microservice is documented at <https://mavlink.io/en/services/ping.html>."]
#[doc = ""]
#[doc = "ID: 4"]
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
#[cfg_attr(feature = "ts", derive(TS))]
#[cfg_attr(feature = "ts", ts(export))]
pub struct PING_DATA {
    #[doc = "Timestamp (UNIX Epoch time or time since system boot). The receiving end can infer timestamp format (since 1.1.1970 or since system boot) by checking for the magnitude of the number."]
    pub time_usec: u64,
    #[doc = "PING sequence"]
    pub seq: u32,
    #[doc = "0: request ping from all receiving systems. If greater than 0: message is a ping response and number is the system id of the requesting system"]
    pub target_system: u8,
    #[doc = "0: request ping from all receiving components. If greater than 0: message is a ping response and number is the component id of the requesting component."]
    pub target_component: u8,
}
impl PING_DATA {
    pub const ENCODED_LEN: usize = 14usize;
    pub const DEFAULT: Self = Self {
        time_usec: 0_u64,
        seq: 0_u32,
        target_system: 0_u8,
        target_component: 0_u8,
    };
    #[cfg(feature = "arbitrary")]
    pub fn random<R: rand::RngCore>(rng: &mut R) -> Self {
        use arbitrary::{Arbitrary, Unstructured};
        let mut buf = [0u8; 1024];
        rng.fill_bytes(&mut buf);
        let mut unstructured = Unstructured::new(&buf);
        Self::arbitrary(&mut unstructured).unwrap_or_default()
    }
}
impl Default for PING_DATA {
    fn default() -> Self {
        Self::DEFAULT.clone()
    }
}
impl MessageData for PING_DATA {
    type Message = MavMessage;
    const ID: u32 = 4u32;
    const NAME: &'static str = "PING";
    const EXTRA_CRC: u8 = 237u8;
    const ENCODED_LEN: usize = 14usize;
    fn deser(
        _version: MavlinkVersion,
        __input: &[u8],
    ) -> Result<Self, ::mavlink_core::error::ParserError> {
        let avail_len = __input.len();
        let mut payload_buf = [0; Self::ENCODED_LEN];
        let mut buf = if avail_len < Self::ENCODED_LEN {
            payload_buf[0..avail_len].copy_from_slice(__input);
            Bytes::new(&payload_buf)
        } else {
            Bytes::new(__input)
        };
        let mut __struct = Self::default();
        __struct.time_usec = buf.get_u64_le();
        __struct.seq = buf.get_u32_le();
        __struct.target_system = buf.get_u8();
        __struct.target_component = buf.get_u8();
        Ok(__struct)
    }
    fn ser(&self, version: MavlinkVersion, bytes: &mut [u8]) -> usize {
        let mut __tmp = BytesMut::new(bytes);
        #[allow(clippy::absurd_extreme_comparisons)]
        #[allow(unused_comparisons)]
        if __tmp.remaining() < Self::ENCODED_LEN {
            panic!(
                "buffer is too small (need {} bytes, but got {})",
                Self::ENCODED_LEN,
                __tmp.remaining(),
            )
        }
        __tmp.put_u64_le(self.time_usec);
        __tmp.put_u32_le(self.seq);
        __tmp.put_u8(self.target_system);
        __tmp.put_u8(self.target_component);
        if matches!(version, MavlinkVersion::V2) {
            let len = __tmp.len();
            ::mavlink_core::utils::remove_trailing_zeroes(&bytes[..len])
        } else {
            __tmp.len()
        }
    }
}
#[derive(Clone, PartialEq, Debug)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type"))]
#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
#[cfg_attr(feature = "ts", derive(TS))]
#[cfg_attr(feature = "ts", ts(export))]
#[repr(u32)]
pub enum MavMessage {
    #[doc = "A ping message either requesting or responding to a ping. This allows to measure the system latencies, including serial port, radio modem and UDP connections. The ping microservice is documented at <https://mavlink.io/en/services/ping.html>."]
    #[doc = ""]
    #[doc = "ID: 4"]
    #[deprecated = "To be removed / merged with TIMESYNC. See `TIMESYNC` (Deprecated since 2011-08)"]
    PING(PING_DATA),
}
impl MavMessage {
    pub const fn all_ids() -> &'static [u32] {
        &[4u32]
    }
}
impl Message for MavMessage {
    fn parse(
        version: MavlinkVersion,
        id: u32,
        payload: &[u8],
    ) -> Result<Self, ::mavlink_core::error::ParserError> {
        match id {
            PING_DATA::ID => PING_DATA::deser(version, payload).map(Self::PING),
            _ => Err(::mavlink_core::error::ParserError::UnknownMessage { id }),
        }
    }
    fn message_name(&self) -> &'static str {
        match self {
            Self::PING(..) => PING_DATA::NAME,
        }
    }
    fn message_id(&self) -> u32 {
        match self {
            Self::PING(..) => PING_DATA::ID,
        }
    }
    fn message_id_from_name(name: &str) -> Option<u32> {
        match name {
            PING_DATA::NAME => Some(PING_DATA::ID),
            _ => None,
        }
    }
    fn default_message_from_id(id: u32) -> Option<Self> {
        match id {
            PING_DATA::ID => Some(Self::PING(PING_DATA::default())),
            _ => None,
        }
    }
    #[cfg(feature = "arbitrary")]
    fn random_message_from_id<R: rand::RngCore>(id: u32, rng: &mut R) -> Option<Self> {
        match id {
            PING_DATA::ID => Some(Self::PING(PING_DATA::random(rng))),
            _ => None,
        }
    }
    fn ser(&self, version: MavlinkVersion, bytes: &mut [u8]) -> usize {
        match self {
            Self::PING(body) => body.ser(version, bytes),
        }
    }
    fn extra_crc(id: u32) -> u8 {
        match id {
            PING_DATA::ID => PING_DATA::EXTRA_CRC,
            _ => 0,
        }
    }
    fn target_system_id(&self) -> Option<u8> {
        match self {
            Self::PING(inner) => Some(inner.target_system),
            _ => None,
        }
    }
    fn target_component_id(&self) -> Option<u8> {
        match self {
            Self::PING(inner) => Some(inner.target_component),
            _ => None,
        }
    }
}
